<img src="http://imgur.com/1ZcRyrc.png" style="float: left; margin: 20px; height: 55px">

# Inferential Statistics Lab

_Author: Matt Brems (DC)_

Head to the `data` folder. There are two files:
- `crx_names.txt`: This summarizes the source of the data and provides information that is important to understanding what the data mean. **Be sure to read this first!**
- `crx_data.csv`: This will be the data itself. **Note that there are no column headers.**

A source of the data is [here](https://archive.ics.uci.edu/ml/datasets/Japanese+Credit+Screening) if you would like to learn more.

**Exercise 1**: Load the data in using any method that you choose.

In [34]:
import pandas as pd

df = pd.read_csv('/Users/kylesokolis/dsi/submissions/data/crx_data.csv', header=None)

In [35]:
df.head()

0      1      2  3  4  5  6     7  8  9   10 11 12     13   14 15
0  b  30.83  0.000  u  g  w  v  1.25  t  t   1  f  g  00202    0  +
1  a  58.67  4.460  u  g  q  h  3.04  t  t   6  f  g  00043  560  +
2  a  24.50  0.500  u  g  q  h  1.50  t  f   0  f  g  00280  824  +
3  b  27.83  1.540  u  g  w  v  3.75  t  t   5  t  g  00100    3  +
4  b  20.17  5.625  u  g  w  v  1.71  t  f   0  f  s  00120    0  +

**Exercise 2**: Note that there are no meaningful column names. Why is this the case? Do you agree with this or disagree with this?

# **Answer**:

# Though I don't necessarily like this, I do understand it's necessity at times. Confidentiality is a very important thing when dealing with people's data.

**Exercise 3:** You want to give names to each column. Read the following line of code:

```python
['X' + str(i) for i in range(1,17)]
```

Before running this line of code, what will this create? Go ahead and use this for the column names.

# **Answer**:

# This will give the name 'X#' consecutively through the columns.

In [36]:
df.columns = ['X' + str(i) for i in range(1,17)]

In [37]:
df.head()

X1     X2     X3 X4 X5 X6 X7    X8 X9 X10  X11 X12 X13    X14  X15 X16
0  b  30.83  0.000  u  g  w  v  1.25  t   t    1   f   g  00202    0   +
1  a  58.67  4.460  u  g  q  h  3.04  t   t    6   f   g  00043  560   +
2  a  24.50  0.500  u  g  q  h  1.50  t   f    0   f   g  00280  824   +
3  b  27.83  1.540  u  g  w  v  3.75  t   t    5   t   g  00100    3   +
4  b  20.17  5.625  u  g  w  v  1.71  t   f    0   f   s  00120    0   +

**Exercise 4**: Count the number of missing values in each column. (There are multiple ways to do this.)

In [38]:
df.isnull().sum()

X1     0
X2     0
X3     0
X4     0
X5     0
X6     0
X7     0
X8     0
X9     0
X10    0
X11    0
X12    0
X13    0
X14    0
X15    0
X16    0
dtype: int64

In [39]:
df['X1'].value_counts()

b    468
a    210
?     12
Name: X1, dtype: int64

# Answer:

# There are 12 missing values, but they have are currently named '?'. Deciding to change '?' to NaN and then dropping NaN rows.

In [40]:
import numpy as np

In [41]:
df.replace('?', np.nan, inplace=True)

In [42]:
df.isnull().sum()

X1     12
X2     12
X3      0
X4      6
X5      6
X6      9
X7      9
X8      0
X9      0
X10     0
X11     0
X12     0
X13     0
X14    13
X15     0
X16     0
dtype: int64

In [43]:
df.shape

(690, 16)

In [44]:
df = df.dropna()

In [33]:
df.isnull().sum()

X1     0
X2     0
X3     0
X4     0
X5     0
X6     0
X7     0
X8     0
X9     0
X10    0
X11    0
X12    0
X13    0
X14    0
X15    0
X16    0
dtype: int64

In [45]:
df.shape

(653, 16)

**Exercise 5**: Our goal is to learn about the population of interest. In this case, our population is all credit applications submitted to this company.

How would you describe the sample here?

# **Answer**:

# The orginal sample was 690 credit applications, though it is now 653 credit applications after dropping NaN rows.

**Exercise 6**: Our goal is to learn about the population of interest. In this case, our population is all credit applications submitted to this company. We specifically want to estimate the true proportion of approved applications for this company.

What is the parameter here? What is the statistic here? 
> Be sure to identify which is which!

# **Answer**: 

# Parameter = True proportion of approved
# Statistic = Sample propotion of approved

**Exercise 7**: Recall that the formula for a confidence interval is:

$$
\text{[sample statistic] } \pm \text{[multiplier] } \times \text{[standard deviation of sampling distribution]}
$$

Calculate the:
- sample percentage of `+` applications
- sample standard deviation of `+` applications
- size of our sample.

Use these to generate a 95% confidence interval for the true proportion of approved applications for this company. Note that column `X16` identifies which applications in our sample were approved (`+`) and denied (`-`).

> Some data "munging" (cleaning/transforming) may be required!

In [62]:
df_new = pd.get_dummies(df['X16'], drop_first=True)

In [ ]:
df = df.join(df_new)

In [67]:
df.head()

X1     X2     X3 X4 X5 X6 X7    X8 X9 X10  X11 X12 X13    X14  X15 X16  -
0  b  30.83  0.000  u  g  w  v  1.25  t   t    1   f   g  00202    0   +  0
1  a  58.67  4.460  u  g  q  h  3.04  t   t    6   f   g  00043  560   +  0
2  a  24.50  0.500  u  g  q  h  1.50  t   f    0   f   g  00280  824   +  0
3  b  27.83  1.540  u  g  w  v  3.75  t   t    5   t   g  00100    3   +  0
4  b  20.17  5.625  u  g  w  v  1.71  t   f    0   f   s  00120    0   +  0

In [71]:
df = df.drop(['X16'], axis=1)

In [73]:
df = df.rename(columns={'-': 'X16'})

In [75]:
df.head()

X1     X2     X3 X4 X5 X6 X7    X8 X9 X10  X11 X12 X13    X14  X15  X16
0  b  30.83  0.000  u  g  w  v  1.25  t   t    1   f   g  00202    0    0
1  a  58.67  4.460  u  g  q  h  3.04  t   t    6   f   g  00043  560    0
2  a  24.50  0.500  u  g  q  h  1.50  t   f    0   f   g  00280  824    0
3  b  27.83  1.540  u  g  w  v  3.75  t   t    5   t   g  00100    3    0
4  b  20.17  5.625  u  g  w  v  1.71  t   f    0   f   s  00120    0    0

In [76]:
df['X16'].value_counts()

1    357
0    296
Name: X16, dtype: int64

In [ ]:
sample_pct = np.mean(df['X16'])
sigma = np.std(df['X16'])
sample_size = len(df['X16'])

In [83]:
(round(sample_pct - 1.96 * sigma / (sample_size ** 0.5), 2)*100)

51.0

In [84]:
(round(sample_pct + 1.96 * sigma / (sample_size ** 0.5), 2)*100)

57.99999999999999

# **Answer**:

# My 95% confidence interval for the true population proportion of approved credit applications is (51.0%, 57.9%).

**Exercise 8**: Interpret the above interval.
> While you _could_ copy and paste text from the notes and fill in the blanks, you should practice interpreting the interval. Remember, this will come up in interviews!

# **Answer**:

# I am 95% confident that the true population proportion of approved credit applications is between 51.0% and 57.9%.

**Exercise 9**: Define a function named `conf_int()` that accepts two arguments: 
- `data`, which should be an array or Series of data
- `conf_level`, which should be either `90`, `95`, or `99`.

Your function should return the 90% confidence interval, 95% confidence interval, or 99% confidence interval, depending on what value the user selected. **Set the default to be 95.**

> For a 90% confidence interval, the multiplier is 1.645.

> For a 95% confidence interval, the multiplier is 1.96.

> For a 99% confidence interval, the multiplier is 2.576.

In [89]:
def conf_int(data, conf_level = 95):
    
    if conf_level == 95:
        multiplier = 1.96
    elif conf_level == 90:
        multiplier = 1.645
    elif conf_level == 99:
        multiplier = 2.576
    else:
        return 'Try Again'
    
    sample_pct = np.mean(data)
    sigma = np.std(data)
    sample_size = len(data)
    
    return ((round(sample_pct - multiplier * sigma / (sample_size ** 0.5), 2)*100),
            (round(sample_pct + multiplier * sigma / (sample_size ** 0.5), 2)*100))

**Exercise 10**: Test your function to find the 99% confidence interval for the mean of `X3`. Your answer should be **(4.2709, 5.2466)**. Also interpret the interval.

In [90]:
print("The 90% confidence interval is: " + str(conf_int(df['X3'], 90)))
print("The 95% confidence interval is: " + str(conf_int(df['X3'], 95)))
print("The 99% confidence interval is: " + str(conf_int(df['X3'], 99)))

The 90% confidence interval is: (451.0, 515.0)
The 95% confidence interval is: (444.00000000000006, 521.0)
The 99% confidence interval is: (432.0, 534.0)


# **Answer**: 

# I am 99% confident that the true average value of $X_3$ is between 432.0 and 534.0.

**Exercise 11**: We want to test whether or not the mean of $X_3$ was equal to 5.

State the null and alternative hypotheses.

# **Answer:**

# H_0: mu = 5
# H_1: mu != 5

**Exercise 12**: Use a one-sample $t$-test to test the above hypotheses at the $\alpha = 0.05$ significance level. Report and interpret your $p$-value.
> Hint: You might find [this link](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_1samp.html) to be helpful! Check out the `a` and `popmean` arguments.

In [92]:
import scipy.stats as stats

In [93]:
stats.ttest_1samp(df['X3'], popmean = 5)

Ttest_1sampResult(statistic=-0.8665258535552368, pvalue=0.3865206185576372)

# **Answer**:

# pvalue = 0.387. 

# (0.387>0.05) 

# There is insufficient evidence to reject H_0 and cannot conclude that the mean of X3 is different from 5.

**Exercise 13**: We want to test whether or not the true proportion of $X_{16}$ was equal to 0.5.

State the null and alternative hypotheses.

# **Answer**:

# H_0: mu = 0.5
# H_1: mu != 0.5

**Exercise 14**: Use a one-sample $t$-test to test the above hypotheses at the $\alpha = 0.05$ significance level. Report and interpret your $p$-value.
> Hint: You might find [this link](https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.ttest_1samp.html) to be helpful! Check out the `a` and `popmean` arguments.

In [94]:
stats.ttest_1samp(df['X16'], 0.5)

Ttest_1sampResult(statistic=2.3957620082073485, pvalue=0.016866793153617506)

# **Answer**:

# pvalue = 0.017

# 0.017 < 0.05 

# I reject H_0. Thus, I conclude that the mean of X16 is different from 0.5.